File: erlang/prime-factors/src/prime_factors.erl
-module(prime_factors).
-export([factors/1]).

factors(Value) -> aux(Value, 2).

aux(1, _) -> [];
aux(Value, Factor) when Value rem Factor == 0 -> [Factor | aux(Value div Factor, Factor)];
aux(Value, Factor) -> aux(Value, Factor + 1).

File: elixir/prime-factors/lib/prime_factors.ex
defmodule PrimeFactors do
  def factors_for(number) when number < 2, do: []

  def factors_for(number) do
    Stream.unfold(
      {number, 2},
      fn
        {1, _} ->
          nil
        {n, d} ->
          d = next_divisor(n, d)
          {d, {div(n, d), d}}
      end
    )
    |> Enum.to_list()
  end

  defp next_divisor(n, d) when rem(n, d) == 0, do: d
  defp next_divisor(n, d) when d * d > n, do: n
  defp next_divisor(n, d), do: next_divisor(n, d + 1)
end
